pycloudlib/lxd/instance.py

```diff
@@ -333,3 +334,82 @@ def wait_for_stop(self):
                 return
             time.sleep(1)
         raise TimeoutError
+
+
+class LXDVirtualMachineInstance(LXDInstance):
+    """LXD Virtual Machine backed instance."""
+
+    def __init__(self, name, key_pair=None, execute_via_ssh=True):
+        """Set up instance.
+
+        Args:
+            name: name of instance
+            key_pair: SSH key object
+        """
+        super().__init__(
+            name=name,
+            key_pair=key_pair,
+            execute_via_ssh=execute_via_ssh
+        )
+
+        self.series = self._extract_release_from_instance_info()
+
+    def _run_command(self, command, stdin):
+        """Run command in the instance."""
+        if self.execute_via_ssh:
+            return super()._run_command(command, stdin)
+
+        if self.series == "xenial":
+            msg = (
+                "The xenial instance does not support executing commads"
+                " with exec."
+            )
+            self._log.warning(msg)
+
+        return super()._run_command(command, stdin)
+
+    def _lxc_instance_info(self) -> dict:
+        """Return a dict of the output of ``lxc info <instance_name>``.
+
+        Args:
+            instance_name: string, name of the LXD instance
+
+        Returns:
+            A dict produced by loading the YAML emitted by ``lxc info
+            <instance_name>``, or the empty dict if either the command
+            or YAML load fails.
+        """
+        raw_info = subp(["lxc", "info", self.name], rcs=())
+        if raw_info.ok:
+            try:
+                return yaml.safe_load(raw_info.replace("\t", " "))
+            except yaml.YAMLError:
+                pass
+        return {}
+
+    def _extract_release_from_instance_info(self):
+        """Extract the base release from the instance.
+
+        We will use the profile created for LXD virtual machines
+        to infer which release was used to create that instance
+
+        Returns:
+            A string containing the base release from the instance
+        """
+        instance_info = self._lxc_instance_info()
+        release = None
+        try:
+            profiles = str(instance_info["Profiles"])
+
+            pattern_search = re.search(
+                r"pycloudlib-vm-(?P<release>.*)-v.*",
```

Comment:
This has been updated for the change in profile naming (thanks!), but this still doesn't address the fact that this will stop working entirely once we fix #75 (i.e. once we have a profile that's used for more than a single Ubuntu release, and therefore won't include a release name in its name). This determination also won't work for instances launched without pycloudlib (or even with pycloudlib, if the consumer specifies their own profile on launch).

Given this, I'd lean towards dropping this: it's quite a lot of code to "just" emit a warning, and it will likely have to be removed or substantially modified when #75 happens.

I think we agreed that you'd switch to just using the image ID for this determination, if we have it available?

Reply:
Got it. I have dropped all of the code that performed that verification and added the `image id` way of checking the release